### PR TITLE
Add Finding of Child Assets

### PR DESCRIPTION
**Pull request by Asura5**
Jira: https://hackney.atlassian.net/browse/MR-708
Proposal doc: https://docs.google.com/document/d/1Wg39ByH4uuroPvFMkuq2TQt4rnqU3DDyPOvoYwytLys/edit

- Nominally, an asset just has a "parentAssetIds" field for navigation between directly related assets
- This, from a CS perspective, is a single-linked list
- The ticket calls for a view over child assets (descendants) as well as antecedents (parent/root assets)
- ElasticSearch indexing allows us to do this, therefore, add finding child assets via a new endpoint

Endpoint use:
`GET /api/v1/search/assetrelationships?searchText=<Asset ID>`

Example:
`GET /api/v1/search/assetrelationships?searchText=a65918aa-e441-bc34-e49d-fa8671f768da`

Returned object: `GetAssetRelationshipsResponse`

The root (level 1) from the original asset
The parents (level 2) from the original asset
The asset itself
The Child objects from this call

### CODE REVIEW

File: HousingSearchApi/V1/Controllers/GetAssetRelationshipController.cs [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    {
+        private readonly IGetAssetRelationshipsUseCase _getAssetRelationshipsUseCase;
+
+        public GetAssetRelationshipController(IGetAssetRelationshipsUseCase getAssetRelationshipsUseCase)
+        {
+            _getAssetRelationshipsUseCase = getAssetRelationshipsUseCase;
+        }
+
+        [ProducesResponseType(typeof(APIResponse<GetAssetRelationshipsResponse>), 200)]
+        [ProducesResponseType(typeof(APIResponse<NotFoundException>), 404)]
```
Comment (LBHCallumM) [inline HousingSearchApi/V1/Controllers/GetAssetRelationshipController.cs:28]: Did you intend to return a NotFoundResponse?

Comment (Asura5) [reply HousingSearchApi/V1/Controllers/GetAssetRelationshipController.cs:28]: Hi mate - I'm in the process of adding validation - it's not ready for review yet 👍

Comment (Asura5) [reply HousingSearchApi/V1/Controllers/GetAssetRelationshipController.cs:28]: Just posted it to give Angelo the likely frontend contract

Comment (LBHCallumM) [reply HousingSearchApi/V1/Controllers/GetAssetRelationshipController.cs]: I had my suspicions.

File: HousingSearchApi/V1/Infrastructure/Factories/AssetQueryGenerator.cs [hunk 2]
```diff
@@ -22,6 +22,15 @@ public class AssetQueryGenerator : IQueryGenerator<QueryableAsset>
 
         public QueryContainer Create<TRequest>(TRequest request, QueryContainerDescriptor<QueryableAsset> q)
         {
+            //Pre-default check for relationships query
+            if (request.GetType() == typeof(GetAssetRelationshipsRequest))
+            {
+                var relationshipRequest = request as GetAssetRelationshipsRequest;
+                return _queryFilterBuilder
+                    .WithExactQuery(relationshipRequest.SearchText, new List<string> { "parentAssetIds" })
```
Comment (LBHCallumM) [inline HousingSearchApi/V1/Infrastructure/Factories/AssetQueryGenerator.cs:30]: Have you tested this? Doesn't the field contain '#' to separate the ids, meaning an exact match wouldn't work?

File: HousingSearchApi/V1/Controllers/GetAssetRelationshipController.cs [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+        {
+            if (string.IsNullOrWhiteSpace(request.SearchText)) return BadRequest("Request searchtext cannot be blank");
+
+            try
+            {
+                var assetsSearchResult = await _getAssetRelationshipsUseCase.ExecuteAsync(request).ConfigureAwait(false);
+
+                if (!assetsSearchResult.ChildAssets.Any()) return new NoContentResult();
+
+                return new OkObjectResult(assetsSearchResult);
```
Comment (LBHCallumM) [inline HousingSearchApi/V1/Controllers/GetAssetRelationshipController.cs:42]: Is there a reason you choose a 204 no content response? From my understanding, I would use 204 if the endpoint never returned anything, such as saving a change to the database. If it was me, I would return a 200 with an empty array instead.

File: HousingSearchApi.Tests/V1/E2ETests/Stories/GetAssetRelationshipsStories.cs [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+        }
+
+        [Fact]
+        public void ServiceReturnsOkResult()
+        {
+            this.Given(g => _assetsFixture.GivenAnAssetIndexExists())
+                .When(w => _steps.WhenRequestContainsSearchString())
+                .Then(t => _steps.ThenTheLastRequestShouldBe200())
+                .BDDfy();
+        }
```
Comment (LBHCallumM) [inline HousingSearchApi.Tests/V1/E2ETests/Stories/GetAssetRelationshipsStories.cs:45]: Should there be a test for a no content response (or no results)?

**Review by LBHCallumM — APPROVED** [review]
Approved, subject to comments

File: HousingSearchApi/V1/Boundary/Requests/GetAssetRelationshipsRequest.cs [hunk 5]
```diff
@@ -0,0 +1,10 @@
+using Microsoft.AspNetCore.Mvc;
+
+namespace HousingSearchApi.V1.Boundary.Requests
+{
+    public class GetAssetRelationshipsRequest : HousingSearchRequest
+    {
+        [FromQuery(Name = "treeLayers")]
+        public int? TreeDepth { get; set; } = 3;
+    }
```
[the remark this answers is not in the document]
Comment (Asura5) [reply HousingSearchApi/V1/Boundary/Requests/GetAssetRelationshipsRequest.cs:9]: tree layers is the number of steps to iterate i.e with a treeLayers value of 2, the children's children would be retrieved. That said, it's a confusing why to implement it I agree, so I'll just stick with direct children, and any grandparenting (if necessary) can come from a recursive call.

[the remark this answers is not in the document]
Comment (Asura5) [reply HousingSearchApi/V1/Boundary/Requests/GetAssetRelationshipsRequest.cs:9]: Changed

File: HousingSearchApi.Tests/V1/E2ETests/Stories/GetAssetRelationshipsStories.cs (hunk 4 above)
Comment (Asura5) [reply HousingSearchApi.Tests/V1/E2ETests/Stories/GetAssetRelationshipsStories.cs:45]: Removed nc response

File: HousingSearchApi/V1/Infrastructure/Factories/AssetQueryGenerator.cs (hunk 2 above)
Comment (Asura5) [reply HousingSearchApi/V1/Infrastructure/Factories/AssetQueryGenerator.cs:30]: I have tested it - and I've included E2E tests to that effect - you misunderstand what an exact query is in ElasticSearch. Say my query is "12 Pitcairn" and the field list is "address1"
Exact query means find "12 Pitcairn" in the field address1 / if a document contains the phrase "12 Pitcairn" anywhere in the field, return the document
Wildstar query means find either 12, or Pitcairn in address1 / if a document contains any of these, return the document

**Pull request merged**